src/merger_tree.cpp

```diff
@@ -0,0 +1,175 @@
+#include <vector>
+#include <cassert>
+
+#include "mpi_wrapper.h"
+#include "merger_tree.h"
+#include "datatypes.h"
+#include "subhalo.h"
+#include "locate_ids.h"
+#include "argsort.h"
+#include "reorder.h"
+
+
+void MergerTreeInfo::Clear() {
+  // Discard the vectors of tracer IDs for the current snapshot
+  DescendantTracerIds.clear();
+}
+
+void MergerTreeInfo::StoreTracerIds(SubhaloList_t &subhalos, int nr_tracers) {
+  // Store tracers for all resolved subhalos
+  for(auto &sub : subhalos) {
+    if(sub.Nbound > 1) {
+      DescendantTracerIds[sub.TrackId] = sub.GetMostBoundTracerIds(nr_tracers);
+    }
+  }
+}
+
+void MergerTreeInfo::FindDescendants(SubhaloList_t &Subhalos, MpiWorker_t world) {
+
+  // Compute descendants for all disrupted subhalos in DescendantTracerIds.
+  // Updates DescendantTrackId in the affected Subhalo_t class instances.
+
+  std::vector<HBTInt> count_found(0);
+  std::vector<HBTInt> trackids_found(0);
+  {
+    // Count particles in resolved halos
+    HBTInt nr_particles = 0;
+    for(auto &subhalo: Subhalos) {
+      if(subhalo.Nbound > 1)nr_particles += subhalo.Nbound;
+    }
+
+    // Populate vectors of IDs of all particles in resolved subhalos
+    std::vector<HBTInt> all_trackids;
+    all_trackids.reserve(nr_particles);
+    std::vector<HBTInt> all_particle_ids;
+    all_particle_ids.reserve(nr_particles);
+    for(const auto &subhalo: Subhalos) {
+      if(subhalo.Nbound > 1) {
+        for(HBTInt i=0; i<subhalo.Nbound; i+=1) {
+          all_trackids.push_back(subhalo.TrackId);
+          all_particle_ids.push_back(subhalo.Particles[i].Id);
+        }
+      }
+    }
+    assert(all_trackids.size()==nr_particles);
+    assert(all_particle_ids.size()==nr_particles);
+  
+    // Count particles from subhalos which became unresolved
```

Comment:
This comment implies that we will be only looking for IDs that belonged to just disrupted objects, but I cannot see how the selection for just disrupted objects occurs. `DescendantTracerIds` is built using information before the unbinding takes place, and there is not extra condition to see if a given entry in `DescendantTracerIds` is associated to a disrupted TrackId.

Do we need to add if statements to the for loops that populate the particleIDs to look for? e.g. `if(Subhalos[subid].SnapshotIndexOfDeath == Snapshot && Subhalos[subid].SnapshotIndexSink == -1)`?

Reply:
If this also meant to account for the descendants of resolved objects (i.e. the same track)?

Reply:
I've updated the comments. I think I'd prefer to compute and output descendants for all resolved objects. It's not particularly expensive and it allows cross checks with TrackId and SinkTrackId for things that survive or sink.